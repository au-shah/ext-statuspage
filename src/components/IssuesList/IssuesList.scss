.flex-container {
    display: grid;
    grid-template-columns: 309px 309px 309px 309px;
    grid-column-gap: 2px;
    align-items: center;
  }

.panel {
    background-color: white;
    width: 309px;
    margin: 1px;
    position: relative;
    height: 165px;
}

.issue-title {
    margin-left: 16px;
    margin-right: 21px;
    font-size: 14px;
    color: #161616;
    letter-spacing: 0.16px;
    line-height: 18px;
}

.div-title {
    overflow: hidden;
    padding-top: 20px;
    width: 1244px;
    height: 30px;
    font-size: 16px;
    color: #161616;
    letter-spacing: 0;
    line-height: 22px;
}

.div-width {
    width: 1244px;
}

.date {
    margin-left: 16px;
    margin-right: 21px;
    font-size: 12px;
    color: #525252;
    letter-spacing: 0.32px;
    text-align: left;
    line-height: 16px;
}

.link {
    text-decoration: none;
    font-size: 14px;
    color: #0F62FE;
    letter-spacing: 0.16px;
    line-height: 18px;
}

.view-issue {
    position: absolute;
    bottom: 16px;
    right: 16px;
}

.button {
    text-decoration: none;
    display: inline-block;
    background: url('../../add.svg');
    background-repeat: no-repeat;
    background-size: 16px;
    background-color: #0f62fe;
    background-position-x: 169px;
    background-position-y: 8px;
    height: 19px;
    width: 185px;
    color: white;
    font-size: 14px;
    padding-top: 3px;
    padding-bottom: 7px;
    padding-left: 16px;
}

.new-issue-position {
    float: right;
    margin-left: 2rem;
}

.view-all-position {
    float: right;
}

.view-milestone {
    float: right;
}
