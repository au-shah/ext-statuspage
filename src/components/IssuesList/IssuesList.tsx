import React, { Component } from 'react';
import './IssuesList.scss';

const githubProject: string = 'IBM-Blockchain/blockchain-vscode-extension';

// Github URLs
const issuesHtmlUrl: string = 'https://github.com/' + githubProject + '/issues?q=';
const newIssueUrl: string = 'https://github.com/' + githubProject + '/issues/new';

interface IssueProps {
    issues?: any;
    error?: any;
    refreshIssues?: any;
    title?: any;
    issueLabel?: any;
    newestMilestone?: any;
    milestoneNumber?: any;
}

class IssuesList extends Component<IssueProps, {}> {
    render(): JSX.Element {
        let issueItems: any = (
            <p className='alert alert-info' key='info'>Please wait, loading status information</p>
        );
        if (this.props.error) {
            issueItems = (
                <p className='alert alert-danger' key='error'>{this.props.error}</p>
            );
        } else if (this.props.issues !== undefined) {
            issueItems = this.props.issues.map((issue: any) => {
                const creationDate: Date = new Date(issue.created_at);
                const updateDate: Date = new Date(issue.updated_at);
                let updateOrCloseDate: string = 'Updated: ' + updateDate.toLocaleDateString();
                if (issue.state === 'closed') {
                    updateOrCloseDate = 'Resolved: ' + updateDate.toLocaleDateString();
                }

                return (
                    <div className='panel'>
                        <div>
                            <p className='issue-title'>
                                {(issue.state === 'closed' ? 'RESOLVED: ' : '') + issue.title + ' (#' + issue.number + ')'}
                            </p>
                        </div>
                        <div>
                            <span className='date'>{'Reported: ' + creationDate.toLocaleDateString()}</span>
                            <br></br>
                            <span className='date'>{updateOrCloseDate}</span>
                        </div>
                        <a className='link view-issue' href={issue.html_url}>{'View on GitHub'}</a>
                    </div>
                );
            });
            if (issueItems.length === 0) {
                if (this.props.newestMilestone) {
                    issueItems = (
                        <p className='alert alert-danger'>
                            {'No completed issues!'}
                        </p>
                    );
                } else {
                    issueItems = (
                        <p className='alert alert-success'>
                            {'No incidents found!'}
                        </p>
                    );
                }
            }
        }

        let milestoneUrl: string = 'https://github.com/' + githubProject + '/milestone/';
        milestoneUrl += this.props.milestoneNumber;
        const milestoneRelease: boolean = (this.props.newestMilestone !== undefined) ? true : false;
        // const seeMilestoneLink: JSX.Element = (this.props.newestMilestone !== undefined) ? (<a className='link' href={milestoneUrl}> {'See current milestone'} </a>) : (<a className='link' href={issuesHtmlUrl}> {'View all on GitHub'}</a>);
        const reportIncident: JSX.Element = (this.props.newestMilestone !== undefined) ? (<></>) : (<a className='button' href={newIssueUrl}>Report a new issue</a>);

        return(
            <div>
                <div className='div-title'>
                    {this.props.title}

                    <span className='new-issue-position'>{reportIncident}</span>
                    {milestoneRelease ? (
                        <span className='view-milestone'>
                            <a className='link' href={milestoneUrl}> {'See current milestone'} </a>
                        </span>
                    ) : (
                        <span className='view-all-position'>
                            <a className='link' href={issuesHtmlUrl}> {'View all on GitHub'} </a>
                        </span>
                    )}
                </div>
                <div className='flex-container'>
                    {issueItems}
                </div>
            </div>
        );
    }
}

export default IssuesList;
