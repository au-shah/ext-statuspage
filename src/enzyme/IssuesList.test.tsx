import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import chai from 'chai';
import sinon, { SinonSandbox } from 'sinon';
import sinonChai from 'sinon-chai';
import IssuesList from '../components/IssuesList/IssuesList';
chai.should();
chai.use(sinonChai);

// tslint:disable: no-unused-expression

describe('IssuesList', () => {
    let mySandBox: sinon.SinonSandbox;

    beforeEach(async () => {
        mySandBox = sinon.createSandbox();
    });

    afterEach(async () => {
        mySandBox.restore();
    });

    it('should render the expected snapshot', async () => {
        const component: any = renderer
            .create(<IssuesList />)
            .toJSON();
        expect(component).toMatchSnapshot();
    });

    it('should show error test if error is passed as a prop', async () => {
        const component: any = await mount(<IssuesList error={'someError'}/>);
        component.text().includes('someError').should.be.true;
    });

    it('should show message if no incidents to show', async () => {
        const component: any = await mount(<IssuesList issues={[]}/>);
        component.text().includes('No incidents found!').should.be.true;
    });

    it('should show message if no completed issues to show from the most recent milestone', async () => {
        const component: any = await mount(<IssuesList issues={[]} newestMilestone={'someMilestone'}/>);
        component.text().includes('No completed issues!').should.be.true;
    });

    it('should show any incidents', async () => {
        const incidents: any = {
            url: 'someURL',
            repositoryurl: 'someOtherURL',
            number: 1946,
            title: 'someIncident issue',
            state: 'open'
        };
        const component: any = await mount(<IssuesList issues={[incidents]}/>);
        component.text().includes('someIncident issue').should.be.true;
    });

    it('should show any completed issues for the next release', async () => {
        const completedIssues: any = [{
            url: 'someURL',
            repositoryurl: 'someOtherURL',
            number: 1932,
            title: 'Some completed issue',
            state: 'closed',
            labels: [{
                id : 'someLabelID',
                name: 'bug',
                description: 'someLabelDescription'
            }],
            milestone: {
                number: 'milestoneNumber',
                id: 'someMilestoneID'
            }
        },
        {
            url: 'someURL1',
            repositoryurl: 'someOtherURL1',
            number: 1922,
            title: 'Some other completed issue',
            state: 'closed',
            labels: [{
                id : 'someOtherLabelID',
                name: 'enhancement',
                description: 'someOtherLabelDescription'
            }],
            milestone: {
                number: 'milestoneNumber',
                id: 'someMilestoneID'
            }
        }];
        const component: any = await mount(<IssuesList issues={completedIssues} newestMilestone={'someMilestone'}/>);
        component.text().includes('Some completed issue').should.be.true;
        component.text().includes('Some other completed issue').should.be.true;
        component.text().includes('bug').should.be.true;
        component.text().includes('feature').should.be.true;
    });
});
