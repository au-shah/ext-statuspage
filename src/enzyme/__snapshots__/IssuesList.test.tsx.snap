// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssuesList should render the expected snapshot 1`] = `
<div>
  <p
    className="div-title"
  >
    <span
      className="new-issue-position"
    >
      <a
        className="button"
        href="https://github.com/IBM-Blockchain/blockchain-vscode-extension/issues/new"
      >
        Report a new issue
      </a>
    </span>
    <a
      className="link"
      href="https://github.com/IBM-Blockchain/blockchain-vscode-extension/issues?q="
    >
       
      View all on GitHub
    </a>
  </p>
  <div
    className="flex-container"
  >
    <p
      className="alert alert-info"
    >
      Please wait, loading status information
    </p>
  </div>
</div>
`;
